pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,413 @@
+	}
+
+	return nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if i != nil {
+		return &alibabaInstance{instance: i}, nil
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.Labels == nil {
+		c.Labels = map[string]string{}
+	}
+
+	c.Labels[machineUIDTag] = string(machine.UID)
+
+	createInstanceRequest := ecs.CreateCreateInstanceRequest()
+	createInstanceRequest.ImageId, _ = getImageIDForOS(pc.OperatingSystem)
+	createInstanceRequest.InstanceName = machine.Name
+	createInstanceRequest.InstanceType = c.InstanceType
+	createInstanceRequest.VSwitchId = c.VSwitchID
+	createInstanceRequest.InternetMaxBandwidthOut = requests.Integer(c.InternetMaxBandwidthOut)
+	encodedUserData := base64.StdEncoding.EncodeToString([]byte(userdata))
+	createInstanceRequest.UserData = encodedUserData
+	createInstanceRequest.SystemDiskCategory = "cloud_efficiency"
+	createInstanceRequest.ZoneId = c.ZoneID
+
+	res, err := client.CreateInstance(createInstanceRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create instance at Alibaba cloud: %v", err)
+	}
+
+	var (
+		foundInstance *ecs.Instance
+		maxRetries    = 10
+		tries         = 0
+	)
+
+	for {
+		foundInstance, err = getInstance(client, machine.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if tries >= maxRetries {
+			return nil, fmt.Errorf("instance %v isn't in a ready state", foundInstance.InstanceId)
+		}
+
+		switch instanceStatus(foundInstance.Status) {
+		case pendingStatus:
+			tries++
+		case stoppedStatus:
+			startRequest := ecs.CreateStartInstanceRequest()
+			startRequest.InstanceId = res.InstanceId
+
+			_, err = client.StartInstance(startRequest)
+			if err != nil {
+				return nil, err
+			}
+		case runningStatus:
+			ipAddress := ecs.CreateAllocatePublicIpAddressRequest()
```

Comment:
Does this require the instance to be running? What cleans up the public IP?

Reply:
Yes, otherwise it will return an error, that the machine is not running thus cannot allocate an IP address. About cleaning up the public IP, once the machine/instance is deleted, the public IP address will be unbound and available again.

Reply:
Yeah but it shouldn't be available, it should be deallocated, just like we allocated it before. It doesn't really matter if the thing is in use, as long as its allocated, you have to pay

Reply:
It will be deallocated once the instance which uses it has been deleted/released. Per VPC we have 2000 IP addresses that we can use and once an IP address is allocated they will be 1999 and once the instance which uses this IP is deleted the available IP addresses will be 2000 again which means it gets unbound and deallocated.